### ISSUE TITLE
False-positive "Non-overlapping equality check" comparing `Enum`s `self` to members

### ISSUE THREAD

**Issue by bersbersbers** [python/mypy#17317]
**Bug Report**

`mypy` does not understand that in an Enum, `self` can be compared to its members.

**To Reproduce**

```python
from enum import Enum, auto

class Glass(Enum):
    HALF_FULL = auto()
    HALF_EMPTY = auto()

    def print(self) -> str:
        if self == self.HALF_FULL:
            print("half full")
            return

        raise ValueError("Glass is not half-full.")

# prints "half full"
Glass.HALF_FULL.print()
```

https://mypy-play.net/?mypy=latest&python=3.12&flags=strict

**Expected Behavior**

No error.

**Actual Behavior**

```
main.py:9: error: Non-overlapping equality check (left operand type: "Glass", right operand type: "auto")  [comparison-overlap]
Found 1 error in 1 file (checked 1 source file)
```

**Your Environment**

- Mypy version used: 1.10.0
- Mypy command-line flags: `--strict`
- Mypy configuration options from `mypy.ini` (and other config files): none
- Python version used: 3.12.3

https://github.com/python/mypy/issues/16327 seems related.

**Comment by TomHRiscure** [comment]
I would like to add an issue I encountered that appears related to this issue. Mypy creates a similar false positive "non-overlapping equality check" when an Enum that is a field of a class is modified by some method that is run on that class. I've added a minimal example below:

**To Reproduce:**

```
from enum import Enum


class SWITCH(Enum):
    ON = 1
    OFF = 0


class SwitchFlipper:
    def __init__(self, enum_value: SWITCH):
        self.enum_value = enum_value

    def flip_switch(self) -> None:
        if self.enum_value == SWITCH.ON:
            self.enum_value = SWITCH.OFF
        else:
            self.enum_value = SWITCH.ON


flipper = SwitchFlipper(SWITCH.OFF)
assert flipper.enum_value == SWITCH.OFF
flipper.flip_switch()
assert flipper.enum_value == SWITCH.ON
```

**Expected Behavior**

No error.

**Actual Behavior**

`error: Non-overlapping equality check (left operand type: "Literal[SWITCH.OFF]", right operand type: "Literal[Switch.ON]")  [comparison-overlap]`

**My Environment:**

    Mypy version used: >1.4.0 (issue does not show up in version 1.3.0 or earlier
    Mypy command-line flags: --strict
    Mypy configuration options from mypy.ini (and other config files): none
    Python version used: 3.8.10

**Comment by hanna-becker** [comment]
+1

**Comment by verhovensky** [comment]
Same for me

```
class TochkaDocumentType:
    BENEFICIARY = 'beneficiary'
    DEAL = 'deal'


@dataclass
class TochkaUploadDocumentEntity:
    doc_type: TochkaDocumentType
    beneficiary_id: str
    deal_id: str | None


  upload_message: TochkaUploadDocumentEntity = message
  if upload_message.doc_type == TochkaDocumentType.DEAL:
```

The error:

```
error: Non-overlapping equality check (left operand type: "TochkaDocumentType", right operand type: "str")  [comparison-overlap]
```

Mypy version

```
mypy 1.13.0 (compiled: yes)
```

**Comment by TomHKeysight** [comment]
@verhovensky I think your issue there is that your TochkaDocumentType does not extend Enum, so mypy is correct in observing that you are comparing an object of type TochkaDocumentType to a field of type str.

**Comment by verhovensky** [comment]
@TomHKeysight 
You are right, my bad

**Comment by sterliakov** [comment]
Please note that the narrowing issue reported by @TomHKeysight and @arthurazs is completely irrelevant to the OP and is tracked in #17537. This ticket itself is a duplicate of #10910 modulo `is` vs `==`.
